pkg/queue/breaker_test.go

```diff
@@ -107,17 +116,22 @@ func TestBreakerLargeCapacityRecover(t *testing.T) {
 }
 
 // Attempts to perform a concurrent request against the specified breaker.
+// Will wait for request to either be performed, enqueued or rejected.
 func (b *Breaker) concurrentRequest() request {
-
-	// There is a brief window between when capacity is released and
-	// when it becomes available to the next request.  We yield here
-	// to reduce the likelihood that we hit that edge case.  E.g.
-	// without yielding `go test ./pkg/queue/breaker.* -count 10000`
-	// will fail about 3 runs.
-	runtime.Gosched()
-
 	r := request{lock: &sync.Mutex{}, accepted: make(chan bool, 1)}
 	r.lock.Lock()
+
+	if len(b.activeRequests) < cap(b.activeRequests) {
+		// Expect request to be performed
+		defer waitForQueue(b.activeRequests, len(b.activeRequests)+1)
+	} else if len(b.pendingRequests) < cap(b.pendingRequests) {
+		// Expect request to be queued
+		defer waitForQueue(b.pendingRequests, len(b.pendingRequests)+1)
+	} else {
+		// Expect request to be rejected
+		defer r.wait()
+	}
+
 	var start sync.WaitGroup
 	start.Add(1)
```

Comment:
Is the `WaitGroup` still needed, given the new polling waiting mechanism you introduced?

Reply:
It deadlocks without them, couldn't figure out why...